### PR TITLE
[Hub Generated] Review request for Microsoft.Capacity to add version preview/2019-04-01

### PR DESCRIPTION
**Pull request by Shads**
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Shads/azure-rest-api-specs/RIRP-BillingPlans...master/).

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json [hunk 1]
```diff
@@ -291,6 +294,12 @@
           },
           {
             "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "name": "$expand","description": "May be used to expand the planInformation.",
```
Comment (ArcturusZhang) [inline specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: A linebreak is missing here

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json [hunk 2]
```diff
@@ -157,7 +157,7 @@
       "post": {
         "summary": "Calculate price for a `ReservationOrder`.",
         "description": "Calculate price for placing a `ReservationOrder`.",
-        "operationId": "ReservationOrder_Calculate",
+        "operationId": "ReservationOrder_CalculatePrice",
```
Comment (ArcturusZhang) [inline specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: Breaking change here

**Review by ArcturusZhang — CHANGES_REQUESTED** [review]
Please see the comments I left, and please wait for ARM team's feedback

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json (hunk 2 above)
Comment (Shads) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: It was not intended to create breaking changes in this PR, we'll roll this back.

**Comment by Shads** [comment]
> Please see the comments I left, and please wait for ARM team's feedback

I have removed the change flagged as breaking, is ARM review still required?

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json [hunk 3]
```diff
@@ -720,6 +736,25 @@
...
+                "$ref": "#/definitions/ReservationTerm",
+                "description": "The term for the billing SKU is available for."
+              }
+            },
+            "additionalProperties": {
+              "type": "array",
+              "items": {
+                "$ref": "#/definitions/ReservationBillingPlan"
+              }
+            }
```
Comment (majastrz) [inline specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: This is basically declaring that this object allows arbitrary array properties to be specified. Are these known ahead of time. Is there a known schema?

Comment (majastrz) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: Also what does the property name represent here?

---
In reply to: [325447456](https://github.com/Azure/azure-rest-api-specs/pull/7206#discussion_r325447456) [](ancestors = 325447456)

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json [hunk 4]
```diff
@@ -813,6 +848,18 @@
         }
       }
     },
+    "ReservationBillingPlan": {
```
Comment (majastrz) [inline specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: >ReservationBillingPlan [](start = 5, length = 22)

nit - Until I saw the definition, it wasn't obvious that this is an enum. I wonder if it should be called `ReservationBillingPlanType` or similar.

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/PurchaseReservationOrder.json [hunk 5]
```diff
@@ -11,6 +11,7 @@
         "reservedResourceType": "VirtualMachines",
         "billingScopeId": "/subscriptions/ed3a1871-612d-abcd-a849-c2542a68be83",
         "term": "P1Y",
+        "billingPlan": "Monthly",
```
Comment (majastrz) [inline specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/PurchaseReservationOrder.json]: >"billingPlan": "Monthly", [](start = 8, length = 25)

What happens if the request does not specify `billingPlan`?

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/CalculateReservationOrder.json [hunk 6]
```diff
@@ -10,6 +10,7 @@
         "reservedResourceType": "VirtualMachines",
         "billingScopeId": "/subscriptions/ed3a1871-612d-abcd-a849-c2542a68be83",
         "term": "P1Y",
+        "billingPlan": "Monthly",
```
Comment (majastrz) [inline specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/CalculateReservationOrder.json]: >"billingPlan": "Monthly", [](start = 8, length = 25)

What happens if you skip this from the request?

**Review by majastrz — COMMENTED** [review]
I added some questions. Can you take a look?

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/CalculateReservationOrder.json (hunk 6 above)
Comment (Shads) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/CalculateReservationOrder.json]: The billing plan will be defaulted to "Upfront" which was standard behavior before this update.

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/PurchaseReservationOrder.json (hunk 5 above)
Comment (Shads) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/PurchaseReservationOrder.json]: The billing plan will be defaulted to "Upfront" which was standard behavior before this update.

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json (hunk 3 above)
Comment (Shads) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: billingPlans contains a array of objects, these objects names can be values from the enum #/definitions/ReservationBillingPlan. E.g. P1Y, P3Y
Each of these objects contains an array that can contain the values from the enum #/definitions/ReservationBillingPlan. E.g. Monthly, Upfront.

In the example for this api this looks like:
"billingPlans": [{
              "P1Y": [
                "Upfront",
                "Monthly"
              ]
            }, {
              "P3Y": [
                "Upfront",
                "Monthly"
              ]
            }
          ]

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/PurchaseReservationOrder.json (hunk 5 above)
Comment (majastrz) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/PurchaseReservationOrder.json]: Great. Thanks for confirming.

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/CalculateReservationOrder.json (hunk 6 above)
Comment (majastrz) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/examples/CalculateReservationOrder.json]: Great. Thanks for confirming.

File: specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json (hunk 3 above)
Comment (majastrz) [reply specification/reservations/resource-manager/Microsoft.Capacity/preview/2019-04-01/reservations.json]: Sounds good.

**Review by majastrz — APPROVED** [review]
Looks good to me.

**Review by ArcturusZhang — APPROVED** [review]

**Pull request merged**